# Install and Import Pandas

In [ ]:
pip install pandas
#install pandas

In [3]:
import pandas as pd
# this code will import library
# as = alisaing. shortname for pandas

### Always import library at the beginning

In [8]:
pd.__version__

'2.2.2'

Creating DataFrame

In [9]:
data = {"calories":[420, 380, 390],
        "duration": [50, 40, 45]}
# this is a dictionary

In [10]:
# Use DatFrame function to convert this dict to DataFrame
df = pd.DataFrame(data)

# assign it to df
#df is shorthand for DataFrame

In [12]:
print (df)

   calories  duration
0       420        50
1       380        40
2       390        45


#Import Dataset

In [15]:
from google.colab import files
uploaded = files.upload()

Saving student.xlsx to student (1).xlsx


student.csv or student xlsx will be in the folder




In [5]:
df = pd.read_excel ('student.xlsx')
#excel format

df = pd.read_csv('student.csv') would be for .csv files

In [18]:
df

id         name  class  mark  gender
0    1     John Deo   Four    75  female
1    2     Max Ruin  Three    85    male
2    3       Arnold  Three    55    male
3    4   Krish Star   Four    60  female
4    5    John Mike   Four    60  female
5    6    Alex John   Four    55    male
6    7  My John Rob  Fifth    78    male
7    8       Asruid   Five    85    male
8    9      Tes Qry    Six    78     NaN
9   10     Big John   Four    55  female
10  11       Ronald    Six    89  female
11  12        Recky    Six    94  female
12  13          Kty  Seven    88  female
13  14         Bigy  Seven    88  female
14  15     Tade Row    NaN    88    male
15  16        Gimmy   Four    88    male
16  17        Tumyu    Six    54    male
17  18        Honny   Five    75    male
18  19        Tinny   Nine    18    male
19  20       Jackly   Nine    65  female
20  21   Babby John   Four    69  female
21  22       Reggid  Seven    55  female
22  23        Herod  Eight    79    male
23  24    Tiddy Now  Seven    78    male
24  25     Giff Tow  Seven    88    male
25  26       Crelea  Seven    79    male
26  27          NaN  Three    81     NaN
27  28    Rojj Base  Seven    86  female
28  29  Tess Played  Seven    55    male
29  30    Reppy Red    Six    79  female
30  31  Marry Toeey   Four    88    male
31  32    Binn Rott  Seven    90  female
32  33    Kenn Rein    Six    96  female
33  34     Gain Toe  Seven    69    male
34  35   Rows Noump    Six    88  female

Eploratory Data Aanalysis (EDA)

In [23]:
df.head ()
# will pull the first 5 rows

id        name  class  mark  gender
0   1    John Deo   Four    75  female
1   2    Max Ruin  Three    85    male
2   3      Arnold  Three    55    male
3   4  Krish Star   Four    60  female
4   5   John Mike   Four    60  female

In [22]:
#df.head(10)
# pulls the first 10 as we have specified how many rows.

id         name  class  mark  gender
0   1     John Deo   Four    75  female
1   2     Max Ruin  Three    85    male
2   3       Arnold  Three    55    male
3   4   Krish Star   Four    60  female
4   5    John Mike   Four    60  female
5   6    Alex John   Four    55    male
6   7  My John Rob  Fifth    78    male
7   8       Asruid   Five    85    male
8   9      Tes Qry    Six    78     NaN
9  10     Big John   Four    55  female

In [24]:
df.tail()
# Will return last five rows by default

id         name  class  mark  gender
30  31  Marry Toeey   Four    88    male
31  32    Binn Rott  Seven    90  female
32  33    Kenn Rein    Six    96  female
33  34     Gain Toe  Seven    69    male
34  35   Rows Noump    Six    88  female

In [25]:
df.tail(10)
# will return the last 10 rows.

id         name  class  mark  gender
25  26       Crelea  Seven    79    male
26  27          NaN  Three    81     NaN
27  28    Rojj Base  Seven    86  female
28  29  Tess Played  Seven    55    male
29  30    Reppy Red    Six    79  female
30  31  Marry Toeey   Four    88    male
31  32    Binn Rott  Seven    90  female
32  33    Kenn Rein    Six    96  female
33  34     Gain Toe  Seven    69    male
34  35   Rows Noump    Six    88  female

When exploring a new data set always check the head and tail

In [28]:
df.sample ()
# Will return one random row of data - each sample will be different each time.

id      name  class  mark gender
1   2  Max Ruin  Three    85   male

In [29]:
df.sample (5)
# will return 5 random rows of data

id         name  class  mark gender
23  24    Tiddy Now  Seven    78   male
14  15     Tade Row    NaN    88   male
2    3       Arnold  Three    55   male
8    9      Tes Qry    Six    78    NaN
28  29  Tess Played  Seven    55   male

In [30]:
df.shape
# number of rows and number of columns.

(35, 5)

In [32]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35 entries, 0 to 34
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   id      35 non-null     int64 
 1   name    34 non-null     object
 2   class   34 non-null     object
 3   mark    35 non-null     int64 
 4   gender  33 non-null     object
dtypes: int64(2), object(3)
memory usage: 1.5+ KB


* There are 35 entries.
* if any of the columns has non-null count less than number of entries, that column has null (missing) values

* name column has 1 missing value
* class also has 1 missing values
* gender has 2 missing values
* We will have to handle missing values, we can drop or try to replace them.

In [33]:
df.columns
# returns name of the columns

Index(['id', 'name', 'class', 'mark', 'gender'], dtype='object')

In [35]:
df.describe ()
# statistical summary of data - only returns numeric data
# if std is very close or bigger than mean, there are outliers

id       mark
count  35.000000  35.000000
mean   18.000000  74.657143
std    10.246951  16.401117
min     1.000000  18.000000
25%     9.500000  62.500000
50%    18.000000  79.000000
75%    26.500000  88.000000
max    35.000000  96.000000

* mean = average
* std= standart deviation
* 25% borderline of data

* in this dataset 25% of the students has their mark equal to or less than 62.5
* 25% is the mid point between lowest and middle
* 50% = median

These are known as the percentile range.

* You need to pay attention to std and mean values. If std is close or bigger than mean, there are outliers. (outliers extreme points in data)
* outliers have a huge effect on mean

In [ ]:
df.isnull().sum()
# Will return the amount of null values.

In [38]:
len(df)
# number of rows.

35

In [41]:
df.size
# number of rows multiplied by number of columns

175

In [42]:
# datatypes of the columns
df.dtypes

id         int64
name      object
class     object
mark       int64
gender    object
dtype: object

In [43]:
# Counts missing values per column.

df.isnull().sum()


id        0
name      1
class     1
mark      0
gender    2
dtype: int64

In [44]:
# Counts duplicate rows.
df.duplicated().sum()

np.int64(0)

#Subsetting Datasets

In [47]:
df['name']
# will return the name and the index.

0        John Deo
1        Max Ruin
2          Arnold
3      Krish Star
4       John Mike
5       Alex John
6     My John Rob
7          Asruid
8         Tes Qry
9        Big John
10         Ronald
11          Recky
12            Kty
13           Bigy
14       Tade Row
15          Gimmy
16          Tumyu
17          Honny
18          Tinny
19         Jackly
20     Babby John
21         Reggid
22          Herod
23      Tiddy Now
24       Giff Tow
25         Crelea
26            NaN
27      Rojj Base
28    Tess Played
29      Reppy Red
30    Marry Toeey
31      Binn Rott
32      Kenn Rein
33       Gain Toe
34     Rows Noump
Name: name, dtype: object

In [48]:
df[['id' , 'name', 'gender', 'mark']]
# will return the ID, Name, Gender and Mark
# this is still a data frame.

id         name  gender  mark
0    1     John Deo  female    75
1    2     Max Ruin    male    85
2    3       Arnold    male    55
3    4   Krish Star  female    60
4    5    John Mike  female    60
5    6    Alex John    male    55
6    7  My John Rob    male    78
7    8       Asruid    male    85
8    9      Tes Qry     NaN    78
9   10     Big John  female    55
10  11       Ronald  female    89
11  12        Recky  female    94
12  13          Kty  female    88
13  14         Bigy  female    88
14  15     Tade Row    male    88
15  16        Gimmy    male    88
16  17        Tumyu    male    54
17  18        Honny    male    75
18  19        Tinny    male    18
19  20       Jackly  female    65
20  21   Babby John  female    69
21  22       Reggid  female    55
22  23        Herod    male    79
23  24    Tiddy Now    male    78
24  25     Giff Tow    male    88
25  26       Crelea    male    79
26  27          NaN     NaN    81
27  28    Rojj Base  female    86
28  29  Tess Played    male    55
29  30    Reppy Red  female    79
30  31  Marry Toeey    male    88
31  32    Binn Rott  female    90
32  33    Kenn Rein  female    96
33  34     Gain Toe    male    69
34  35   Rows Noump  female    88

* first [] is for slicing, second [] is for list object.

I can apply different method to subsetted df

In [49]:
df["name"].head()
# will return the first 5 names.

0      John Deo
1      Max Ruin
2        Arnold
3    Krish Star
4     John Mike
Name: name, dtype: object

In [50]:
df["mark"].max()
this will return the maximum mark achievable

96

In [51]:
df['mark'].mean()

np.float64(74.65714285714286)

In [53]:
df['mark'].median()

79.0

In [55]:
df[['id' , 'name', 'gender', 'mark']].sample(5)
# will return 5 random entries including the slicing and list that I specified.

id       name  gender  mark
17  18      Honny    male    75
23  24  Tiddy Now    male    78
10  11     Ronald  female    89
26  27        NaN     NaN    81
5    6  Alex John    male    55

In [56]:
df[['id' , 'name', 'gender', 'mark']][5:10]
# This will return all of the entries between index 5 and 10

id         name  gender  mark
5   6    Alex John    male    55
6   7  My John Rob    male    78
7   8       Asruid    male    85
8   9      Tes Qry     NaN    78
9  10     Big John  female    55

In [57]:
df.loc [0, "name"]
#returns the name from the index 0

'John Deo'

In [ ]:
#return first 3 rows of name gender columns

#df.loc [rows index, column names]

In [ ]:
# df.loc[rows , column names]
df.loc[0:2, ["name", "gender"]]


In [58]:
df.loc[0:4, ['id', 'name']]
# will return the first five rows with the student id and student name
# if you omit the starting index it will begin at 0. Similarly if you omit the
# stopping index then it will continue to the end.

id        name
0   1    John Deo
1   2    Max Ruin
2   3      Arnold
3   4  Krish Star
4   5   John Mike

In [59]:
df.iloc[0:5]
# will return the first 5 results based on a range of indexes

id        name  class  mark  gender
0   1    John Deo   Four    75  female
1   2    Max Ruin  Three    85    male
2   3      Arnold  Three    55    male
3   4  Krish Star   Four    60  female
4   5   John Mike   Four    60  female

* id is at index 0
* name is at index 1
* class is at index 2
* mark is at index 3
* gender is at index 4

In [60]:
df.loc[0:4, ['id', 'name']]
# 4 is included

id        name
0   1    John Deo
1   2    Max Ruin
2   3      Arnold
3   4  Krish Star
4   5   John Mike

In [61]:
df.iloc[0:5, [0, 1]]
# index for column
# 5 is not included

id        name
0   1    John Deo
1   2    Max Ruin
2   3      Arnold
3   4  Krish Star
4   5   John Mike

In [6]:
# How to return name, class and gender of first 5 students using iloc method
df.iloc[0:5, [1, 2, 4]]
# in the range 0:5, 5 is not in the results

name  class  gender
0    John Deo   Four  female
1    Max Ruin  Three    male
2      Arnold  Three    male
3  Krish Star   Four  female
4   John Mike   Four  female

In [7]:
# How to return name, class and gender of first 5 students using loc method
df.loc[0:4, ["name", "class", "gender"]]
# in the range 0:4, 4 is included in the results

name  class  gender
0    John Deo   Four  female
1    Max Ruin  Three    male
2      Arnold  Three    male
3  Krish Star   Four  female
4   John Mike   Four  female

#Filtering Data

In [14]:
df["mark"]>90
# Only returns the boolean value

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11     True
12    False
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20    False
21    False
22    False
23    False
24    False
25    False
26    False
27    False
28    False
29    False
30    False
31    False
32     True
33    False
34    False
Name: mark, dtype: bool

In [8]:
# df[condition]
# Mark greater than 90
df[df["mark"]>90]
# this code will return only true

id       name class  mark  gender
11  12      Recky   Six    94  female
32  33  Kenn Rein   Six    96  female

In [15]:
# Return female students only
df[df["gender"]=="female"]

id        name  class  mark  gender
0    1    John Deo   Four    75  female
3    4  Krish Star   Four    60  female
4    5   John Mike   Four    60  female
9   10    Big John   Four    55  female
10  11      Ronald    Six    89  female
11  12       Recky    Six    94  female
12  13         Kty  Seven    88  female
13  14        Bigy  Seven    88  female
19  20      Jackly   Nine    65  female
20  21  Babby John   Four    69  female
21  22      Reggid  Seven    55  female
27  28   Rojj Base  Seven    86  female
29  30   Reppy Red    Six    79  female
31  32   Binn Rott  Seven    90  female
32  33   Kenn Rein    Six    96  female
34  35  Rows Noump    Six    88  female

In [16]:
df.head()

id        name  class  mark  gender
0   1    John Deo   Four    75  female
1   2    Max Ruin  Three    85    male
2   3      Arnold  Three    55    male
3   4  Krish Star   Four    60  female
4   5   John Mike   Four    60  female

In [17]:
df[df["class"]=="Four"]
# Returns all the reults for students in class 4

id         name class  mark  gender
0    1     John Deo  Four    75  female
3    4   Krish Star  Four    60  female
4    5    John Mike  Four    60  female
5    6    Alex John  Four    55    male
9   10     Big John  Four    55  female
15  16        Gimmy  Four    88    male
20  21   Babby John  Four    69  female
30  31  Marry Toeey  Four    88    male

In [19]:
df[df["class"]=="Four"][['name', 'class', 'gender']]
 #Using both filtering *df[df["class"]=="Four"]* and slicing [['name', 'class', 'gender']]
 # returns only name, class and gender for students in class 4.

name class  gender
0      John Deo  Four  female
3    Krish Star  Four  female
4     John Mike  Four  female
5     Alex John  Four    male
9      Big John  Four  female
15        Gimmy  Four    male
20   Babby John  Four  female
30  Marry Toeey  Four    male

In [20]:
df.loc[df["class"]=="Four", ["name", "class", "gender"]]
# same as above but using the loc function.

name class  gender
0      John Deo  Four  female
3    Krish Star  Four  female
4     John Mike  Four  female
5     Alex John  Four    male
9      Big John  Four  female
15        Gimmy  Four    male
20   Babby John  Four  female
30  Marry Toeey  Four    male

In [33]:
df.iloc[df['class']=="Four",[1, 2, 4]]
# iloc doesnt support filtering.

NotImplementedError: iLocation based boolean indexing on an integer type is not available

* when using multiple columns to filter, enclose each condtion by () and seperate them by &

In [27]:
df[(df["class"] =="Four") & (df['gender'] =="male")]

id         name class  mark gender
5    6    Alex John  Four    55   male
15  16        Gimmy  Four    88   male
30  31  Marry Toeey  Four    88   male

In [32]:
df[(df["mark"]<70) & (df["gender"]=="female")][['name', 'class']]
# returns the name and class of females with marks less than 70

name  class
3   Krish Star   Four
4    John Mike   Four
9     Big John   Four
19      Jackly   Nine
20  Babby John   Four
21      Reggid  Seven

In [35]:
df.loc[(df["mark"]<70) & (df["gender"]=="female"),["name", "class"]]
# Same thing as above but using loc operator.

name  class
3   Krish Star   Four
4    John Mike   Four
9     Big John   Four
19      Jackly   Nine
20  Babby John   Four
21      Reggid  Seven

#Sorting results

* There are always the natural order of the rows. In this instance student info is sorted by id.

In [37]:
df.head ()

id        name  class  mark  gender
0   1    John Deo   Four    75  female
1   2    Max Ruin  Three    85    male
2   3      Arnold  Three    55    male
3   4  Krish Star   Four    60  female
4   5   John Mike   Four    60  female

In [36]:
df.sort_values(by="name")
# sorts the names in alphabetical order

id         name  class  mark  gender
5    6    Alex John   Four    55    male
2    3       Arnold  Three    55    male
7    8       Asruid   Five    85    male
20  21   Babby John   Four    69  female
9   10     Big John   Four    55  female
13  14         Bigy  Seven    88  female
31  32    Binn Rott  Seven    90  female
25  26       Crelea  Seven    79    male
33  34     Gain Toe  Seven    69    male
24  25     Giff Tow  Seven    88    male
15  16        Gimmy   Four    88    male
22  23        Herod  Eight    79    male
17  18        Honny   Five    75    male
19  20       Jackly   Nine    65  female
0    1     John Deo   Four    75  female
4    5    John Mike   Four    60  female
32  33    Kenn Rein    Six    96  female
3    4   Krish Star   Four    60  female
12  13          Kty  Seven    88  female
30  31  Marry Toeey   Four    88    male
1    2     Max Ruin  Three    85    male
6    7  My John Rob  Fifth    78    male
11  12        Recky    Six    94  female
21  22       Reggid  Seven    55  female
29  30    Reppy Red    Six    79  female
27  28    Rojj Base  Seven    86  female
10  11       Ronald    Six    89  female
34  35   Rows Noump    Six    88  female
14  15     Tade Row    NaN    88    male
8    9      Tes Qry    Six    78     NaN
28  29  Tess Played  Seven    55    male
23  24    Tiddy Now  Seven    78    male
18  19        Tinny   Nine    18    male
16  17        Tumyu    Six    54    male
26  27          NaN  Three    81     NaN

In [39]:
df.sort_values(by="mark").head()
# if sorted by numeric value it goes in ascending order. because of .head()
# will return the first 5 results in numerical order ascending.

id         name  class  mark gender
18  19        Tinny   Nine    18   male
16  17        Tumyu    Six    54   male
2    3       Arnold  Three    55   male
5    6    Alex John   Four    55   male
28  29  Tess Played  Seven    55   male

In [43]:
df.sort_values(by="name", ascending=False).head(10)
# descending order
# ascending false will reverse the order. its in descending order

id         name  class  mark  gender
16  17        Tumyu    Six    54    male
18  19        Tinny   Nine    18    male
23  24    Tiddy Now  Seven    78    male
28  29  Tess Played  Seven    55    male
8    9      Tes Qry    Six    78     NaN
14  15     Tade Row    NaN    88    male
34  35   Rows Noump    Six    88  female
10  11       Ronald    Six    89  female
27  28    Rojj Base  Seven    86  female
29  30    Reppy Red    Six    79  female

In [45]:
df.sort_values(by="mark", ascending=False).head(10)
# ascending order

id         name  class  mark  gender
32  33    Kenn Rein    Six    96  female
11  12        Recky    Six    94  female
31  32    Binn Rott  Seven    90  female
10  11       Ronald    Six    89  female
30  31  Marry Toeey   Four    88    male
34  35   Rows Noump    Six    88  female
24  25     Giff Tow  Seven    88    male
14  15     Tade Row    NaN    88    male
15  16        Gimmy   Four    88    male
12  13          Kty  Seven    88  female

In [46]:
# applyting sort after slicing
df[df["class"]=="Six"].sort_values(by="mark", ascending=False).head(10)

id        name class  mark  gender
32  33   Kenn Rein   Six    96  female
11  12       Recky   Six    94  female
10  11      Ronald   Six    89  female
34  35  Rows Noump   Six    88  female
29  30   Reppy Red   Six    79  female
8    9     Tes Qry   Six    78     NaN
16  17       Tumyu   Six    54    male

In [47]:
df.loc[df["class"]=="Six", :].sort_values(by="mark")
# same as above using df.loc function

id        name class  mark  gender
16  17       Tumyu   Six    54    male
8    9     Tes Qry   Six    78     NaN
29  30   Reppy Red   Six    79  female
34  35  Rows Noump   Six    88  female
10  11      Ronald   Six    89  female
11  12       Recky   Six    94  female
32  33   Kenn Rein   Six    96  female

In [51]:
df[(df["gender"] == "female") & (df["mark"] > df["mark"].mean())].sort_values(by="mark", ascending=False)
# returns female students that have a mark higher than the mean average.

id        name  class  mark  gender
32  33   Kenn Rein    Six    96  female
11  12       Recky    Six    94  female
31  32   Binn Rott  Seven    90  female
10  11      Ronald    Six    89  female
34  35  Rows Noump    Six    88  female
12  13         Kty  Seven    88  female
13  14        Bigy  Seven    88  female
27  28   Rojj Base  Seven    86  female
29  30   Reppy Red    Six    79  female
0    1    John Deo   Four    75  female

## value_counts ()

In [52]:
## How many male and female students are there?
df["gender"].value_counts()

gender
male      17
female    16
Name: count, dtype: int64

In [53]:
df['class'].value_counts()
# how many students in each class.

class
Seven    10
Four      8
Six       7
Three     3
Nine      2
Five      2
Fifth     1
Eight     1
Name: count, dtype: int64

Pandas brackets

- [] slicing

- [[]] list of the columns

- () is separating filtering conditions

- .mean() this is function

- .head() this is a method

- .shape this is an attribute

Python Brackets

- [] list

- () tuple

- {} set or dictionary